src/loader.rs

```diff
@@ -11,11 +16,243 @@ impl Loader {
     }
   }
 
-  pub(crate) fn load<'src>(&'src self, path: &Path) -> RunResult<&'src str> {
+  pub(crate) fn load_without_includes<'src>(&'src self, path: &Path) -> RunResult<&'src str> {
     let src = fs::read_to_string(path).map_err(|io_error| Error::Load {
       path: path.to_owned(),
       io_error,
     })?;
     Ok(self.arena.alloc(src))
   }
+
+  pub(crate) fn load_with_includes<'src>(&'src self, path: &Path) -> RunResult<&'src str> {
+    let src = self.perform_load(path, HashSet::new())?;
+    Ok(self.arena.alloc(src))
+  }
+
+  fn perform_load(&self, path: &Path, seen_paths: HashSet<PathBuf>) -> RunResult<String> {
+    let src = fs::read_to_string(path).map_err(|io_error| Error::Load {
+      path: path.to_owned(),
+      io_error,
+    })?;
+
+    self.process_includes(src, path, seen_paths)
+  }
+
+  /// Given the original contents of a Justfile (with include directives), load all the included
+  /// paths to produce one String with the contents of all the files concatenated.
+  ///
+  /// Note that this does not do any parsing yet (i.e. nothing stops a justfile from including a
+  /// file that is not a valid justfile), and that (currently) line numbers in error messages will
+  /// reference lines in this concatenated String rather than probably-more-useful information
+  /// about the original file an error came from.
+  ///
+  /// !include directives are allowed to appear in a justfile only before the first non-blank,
+  /// non-comment line. This is because this processing step occurs before parsing, and we don't
+  /// want to accidentally process the string "!include" when it happens to occur within a string
+  /// or in another post-parsing syntactic context where it shouldn't be.
+  fn process_includes(
```

Comment:
This is only called at the end of `load_with_includes_recursive`, which is itself short, so I'd just merge these two functions.